graph/telemetry/istio/appender/extensions.go

```diff
@@ -0,0 +1,366 @@
+		destName := string(lDestName)
+
+		flags := ""
+		if isRequests || protocol == graph.TCP.Name {
+			lFlags, flagsOk := m["flags"]
+			// "flags" is optional in the TS, if not supplied just leave it empty
+			if flagsOk {
+				flags = string(lFlags)
+			}
+		}
+
+		var code string
+		if isRequests {
+			lProtocol, protocolOk := m["protocol"]
+			lCode, codeOk := m["status_code"]
+
+			if !protocolOk || !codeOk {
+				log.Warningf("Skipping %s, missing expected request TS labels", m.String())
+				continue
+			}
+
+			protocol = string(lProtocol)
+			code = string(lCode)
+
+			if skipRequestsGrpc && protocol == graph.GRPC.Name || skipRequestsHttp && protocol == graph.HTTP.Name {
+				continue
+			}
+		}
+
+		a.addTraffic(ext, trafficMap, metric, val, protocol, code, flags, sourceCluster, sourceIsRoot, sourceNs, sourceName, destCluster, destNs, destName)
+	}
+}
+
+func (a ExtensionsAppender) addTraffic(ext config.ExtensionConfig, trafficMap graph.TrafficMap, metric string, val float64, protocol, code, flags, sourceCluster, sourceIsRoot, sourceNs, sourceName, destCluster, destNs, destName string) {
+	isRoot := sourceIsRoot == "true"
+	source, _, err := a.addNode(ext, trafficMap, isRoot, sourceCluster, sourceNs, sourceName, a.GraphType)
+	if err != nil {
+		log.Warningf("Skipping extension addTraffic (source) in extension, %s", err)
+		return
+	}
+	dest, _, err := a.addNode(ext, trafficMap, false, destCluster, destNs, destName, a.GraphType)
+	if err != nil {
+		log.Warningf("Skipping extension addTraffic (dest), %s", err)
+		return
+	}
+
+	if isRoot {
+		url := a.getUrl(ext, source)
+		if url != urlNotFound {
+			source.Metadata[graph.IsExtension] = &graph.ExtInfo{
+				URL:  url,
+				Name: ext.Name,
+			}
+			dest.Metadata[graph.IsExtension] = &graph.ExtInfo{
+				URL:  url,
+				Name: ext.Name,
+			}
+		}
+	}
+
+	// Extensions may generate duplicate metrics by reporting from both the source and destination. To avoid
+	// processing the same information twice we keep track of the time series applied to a particular edge. The
+	// edgeTSHash incorporates information about the time series' source, destination and metric information,
+	// and uses that unique TS has to protect against applying the same intomation twice.
+	edgeTSHash := fmt.Sprintf("%x", md5.Sum([]byte(fmt.Sprintf("%s:%s:%s:%s:%s:%s", metric, source.Metadata[tsHash], dest.Metadata[tsHash], code, flags, destName))))
+
+	a.addEdgeTraffic(val, protocol, code, flags, destName, source, dest, edgeTSHash)
+}
+
+func (a ExtensionsAppender) addNode(ext config.ExtensionConfig, trafficMap graph.TrafficMap, isRoot bool, cluster, namespace, name, graphType string) (*graph.Node, bool, error) {
+	id, nodeType, err := graph.Id(cluster, namespace, name, "", "", "", "", graphType)
+	if err != nil {
+		return nil, false, err
+	}
+	log.Infof("addNode id=%s, nodeType=%s", id, nodeType) // todo: remove
+	node, found := trafficMap[id]
+	if !found {
+		log.Infof("not found") // todo: remove
```

Comment:
Pro tip - rather than having to comment "//todo:remove" everywhere, just do a `log.Errorf("JAY!! ...the rest of message`). Then just search for "JAY" in the code when you want to delete everythng. This is error level so you don't have to worry about the log level, as error should always be spit out. And this has the added bonus of you quickly seeing your own debug messages in the output.. you'll see "JAY!" on the lines you care about.

You're welcome :)

Reply:
Got it, // JAY!! TODO!!